## Cleaning the data


In [86]:
import pandas as pd 

In [87]:
df = pd.read_csv("data/sales_data.csv",quoting=3)

In [88]:
df = df.dropna(subset=["quantity"])


In [89]:
df = df[df["quantity"] >= 0]

In [90]:
df["unit_price"] = pd.to_numeric(df["unit_price"], errors="coerce")

In [91]:
df = df.dropna(subset=["unit_price"])

In [92]:
df = df.drop_duplicates()

In [93]:
df.columns = df.columns.str.strip('"')

In [95]:
df.iloc[:, 0] = df.iloc[:, 0].str.strip('"')

In [96]:
df["date"]  = pd.to_datetime(df["date"], format="mixed")

In [97]:
df["category"] = df["category"].str.strip().str.title()

In [98]:
df["store"] = df["store"].str.strip().str.title()

In [99]:
print("DATA  QUALITY REPORT")
print("Starting rows:", 600)
print("Row removed - missing quantity:", 1)
print("Row removed - invalid price:", 1)
print("Row removed - negative quantity:", 1)
print("Row removed - duplicate:", 1)
print("Total rows removed:", 4)
print("Clean rows remaining:", len(df))

DATA  QUALITY REPORT
Starting rows: 600
Row removed - missing quantity: 1
Row removed - invalid price: 1
Row removed - negative quantity: 1
Row removed - duplicate: 1
Total rows removed: 4
Clean rows remaining: 596


## Descriptive Analysis

In [100]:
df["revenue"] = df["quantity"] * df["unit_price"]

In [ ]:
print(f" The total revenue generated across all stores: {df['revenue'].sum()}")

## The total revenue generated across all stores during the period was R324303.This represents the combined value of all sales transactions after cleaning the data

 The total revenue generated across all stores: 324303.0


In [ ]:
df.groupby("category")["revenue"].sum().sort_values(ascending=False)

## Beverages generated the highest revenue at R109320 followed by Meat at R104055.Bakery generated the lowest revenue at R22518.This shows that revenue was concentrated more heavily in the Beverages and Meat categories than in Produce and Bakery.

category
Beverages    109320.0
Meat         104055.0
Dairy         65604.0
Produce       22806.0
Bakery        22518.0
Name: revenue, dtype: float64

In [ ]:
df.groupby("store")["revenue"].sum().sort_values(ascending=False)

## Sandton generated the highest revenue at R89137.50, while Rosebank generated the lowest at R67184.50. This shows that the stores did not generate the same amount of revenue during the period

store
Sandton     89137.5
Pretoria    84412.0
Soweto      83569.0
Rosebank    67184.5
Name: revenue, dtype: float64

In [ ]:
df.groupby("product_name")["quantity"].sum().sort_values(ascending=False)

## Rooibos Tea 100g was the best-selling producct by quantity, with 570 units sold.This means it had the highest sales volume among the products in the dataset.

product_name
Rooibos Tea 100g       570.0
Full Cream Milk 2L     566.0
Cheddar Cheese 500g    488.0
Bananas 1kg            464.0
Croissant              452.0
Chicken Breasts 1kg    443.0
Whole Wheat Bread      441.0
Coffee Beans 1kg       432.0
Tomatoes 1kg           396.0
Beef Mince 1kg         377.0
Name: quantity, dtype: float64

In [ ]:
df.groupby("product_name")["revenue"].sum().sort_values(ascending=False)

## Coffee Beans 1kg generated the highest revenue at R77760, This means it contributed more revenue than any other individual product in the dataset.

product_name
Coffee Beans 1kg       77760.0
Chicken Breasts 1kg    53160.0
Beef Mince 1kg         50895.0
Cheddar Cheese 500g    46360.0
Rooibos Tea 100g       31350.0
Full Cream Milk 2L     19244.0
Whole Wheat Bread      12348.0
Tomatoes 1kg           11880.0
Bananas 1kg            11136.0
Croissant              10170.0
Name: revenue, dtype: float64

In [ ]:
## Rooibos Tea = highest quantity
## Coffee Beans = highest revenue
## The product sold in the largest quantity was not the product generating the most revenue.This suggests that sales volume and revenue are influenced by both the number of units sold and the price per unit.

In [113]:
df["month"] = df["date"].dt.to_period("M")

In [114]:
df.groupby("month")["revenue"].sum()

month
2026-01    44140.0
2026-02    47411.5
2026-03    37588.0
2026-04    57409.0
2026-05    77000.0
2026-06    58026.5
2026-07     2728.0
Freq: M, Name: revenue, dtype: float64

In [ ]:
## Revenue flactuated throughout the period.It increased from January to February, decreased in March, then increased substantially through April and reached its highest point in May at R77000. Revenue decreased in June and dropped shaply in July to R2728.

In [ ]:
df["revenue"].mean()

## The average transaction value was approximately R544.13. This means that, across the cleaned dataset, the average transaction generated about R544.13 in revenue.

np.float64(544.1325503355705)

In [ ]:
df["payment_method"].value_counts()

## Mobile was the most commonly used payment method, with 166 transactions. This means more transactions were paid using Mobile than EFT, Cash, or Card in this dataset.

payment_method
Mobile"    166
EFT"       144
Cash"      143
Card"      143
Name: count, dtype: int64

## Critical-Thinking Answers

1. The top product by quantity is Rooibos Tea 100g, while the top product by revenue is Coffee Beans 1kg.This shows that selling the highest number of units does not necessarily generate the highest revenue.Revenue is affected by both the volume of units sold and the price per unit.

In [119]:
df.groupby("category")["quantity"].sum()

category
Bakery        893.0
Beverages    1009.0
Dairy        1054.0
Meat          820.0
Produce       853.0
Name: quantity, dtype: float64

In [124]:
df.groupby("category")["unit_price"].mean()

category
Bakery        25.340164
Beverages    108.512397
Dairy         63.801527
Meat         127.570093
Produce       26.817391
Name: unit_price, dtype: float64

In [123]:
df.groupby("category")["revenue"].sum()

category
Bakery        22518.0
Beverages    109320.0
Dairy         65604.0
Meat         104055.0
Produce       22806.0
Name: revenue, dtype: float64

2. The highest revenue was from beverages at R109320.It did not sell the most units (1009 units, compared to Dairy’s 1054 units) nor did it have the highest unit price (R108,51, compared to Meat’s R127,57), but the combination of sales volume and relatively high prices resulted in the highest total revenue.

3. Monthly revenue is uneven and flactuates throughout the period.Revenue increased from R44140 in January to R47411 in February, then decreased to R37588 in March.It increased again in April and reached its highest point in May at R77000, before falling in June to R58026,50.July shows a sharp decrease to R2728.This unusually low figure may be because only part of July's transactions were recorded.The business should investigate whether the July data is complete before using it as assess monthly performance

4. I would recommend that head office consider investing in Rosebank because it generated the lowest revenue of the four stores, at approximately R67184,50.Investing in a lower-prforming store could provide an opportunity to improve its performance and bring it closer to the other stores.However, I would not make a final investment decision based on the revenue alone.I would first want additional data such as operating costs, profit, number of customers, average customer spending, and store capacity.This would help determine whether Rosebank has genuine potential for improvement and whether an investment would be worthwhile.

5. One limitation of this analysis is that it does not have costs and expenses, so I cannot calculate the profit each store makes.